### PR TITLE
allow mounted devices when using custom install handler

### PR DESCRIPTION
**Pull request by Villemoes**
This is an alternative and somewhat less intrusive approach to solving the problem that #186 is also trying to solve. Instead of having RAUC unmount and mount, this simply allows the update to proceed, provided the user takes full responsibility by providing his own install hook. 

For our use case, that actually turns out to be even more convenient, since we can then more easily preserve some of the U-boot environment that would otherwise be lost when one wipes the boot partition completely.

It still requires opt-in per slot in system.conf, since we don't want to effectively disable the mounted-sanity check for those existing users that just happen to provide their own install hook.

### CODE REVIEW

File: docs/reference.rst [hunk 1]
```diff
@@ -199,6 +199,11 @@ hierarchical separator.
   active slots, so that the inactive one can be selected as the update target.
   The parent slot is referenced using the form ``<slot-class>.<idx>``.
 
+``maybe-mounted``
+  Setting this entry to ``true`` tells RAUC to that the slot
```
Comment (ejoerns) [inline docs/reference.rst]: remove 'to'

File: docs/reference.rst [hunk 2]
```diff
@@ -199,6 +199,11 @@ hierarchical separator.
...
+  may be updated even if it is already mounted. Such a slot can only
+  by updated by a custom install hook.
```
Comment (ejoerns) [inline docs/reference.rst]: s/by/be

**Comment by Villemoes** [comment]
Thanks. Yes, minimally invasive was what I tried to do. I've updated the PR according to your feedback.

Regarding updating the bootloader: First, I don't think VFAT actually supports atomic rename (the VFS may make it appear so when the system is running, but that doesn't help in the case of a crash or powerloss mid-update). Second, we're not just updating a single file - there's also update of the U-boot environment, the boot script, and probably other things as well (perhaps an SPL, and in the case of RPI, lots of firmware files). So we're not pretending that this is a "safe" operation. If I/we figure out a way to actually make this atomic/safe given the limitations of the hardware/firmware/file systems involved, then it makes sense to try to lift that logic into RAUC.

**Comment by Villemoes** [comment]
Can I do anything more to move this forward? We'd really like to see this merged.

**Comment by ejoerns** [comment]
Unfortunately, we had a lot of pain for making the tests stable and green while moving to (more reproducible) docker-based custom test environments and fix some remaining known issues (see #323, #350, #354, #355, #351) that we all wanted to have ready for the 1.0 release.
Most of the work is done now, and as soon as 1.0 is out, we will be able to discuss and merge new pull requests again. Thus please apologize for being a bit irresponsive for new PR's at the moment.

**Comment by Villemoes** [comment]
Can I ask someone to (re)review this? This could close two PRs at once :)

File: src/install.c [hunk 3]
```diff
@@ -879,10 +879,16 @@ static gboolean launch_and_wait_default_handler(RaucInstallArgs *args, gchar* bu
 			goto out;
 		}
 
-		if (dest_slot->mount_point || dest_slot->ext_mount_point) {
+		if ((dest_slot->mount_point || dest_slot->ext_mount_point) &&
+		    !(dest_slot->maybe_mounted && mfimage->hooks.install)) {
+			if (dest_slot->maybe_mounted) {
+				g_set_error(error, R_INSTALL_ERROR, R_INSTALL_ERROR_MOUNTED,
+						"Already mounted device '%s' may only be updated by custom install hook", dest_slot->device);
```
[the remark this answers is not in the document]
Comment (Villemoes) [reply src/install.c]: I don't think so, but I'm not a native speaker. Simply dropping "Already" could also be an option. And yes, "a" is missing.

File: docs/reference.rst [hunk 4]
```diff
@@ -209,6 +209,11 @@ hierarchical separator.
   active slots, so that the inactive one can be selected as the update target.
   The parent slot is referenced using the form ``<slot-class>.<idx>``.
 
+``maybe-mounted=<true/false>``
```
[the remark this answers is not in the document]
Comment (Villemoes) [reply docs/reference.rst]: I alluded to that in the commit message, you can read it either way. And I'm fine with may-be-mounted, or allow-mounted, or anything else that is acceptable.

File: src/install.c [hunk 5]
```diff
@@ -879,10 +879,16 @@ static gboolean launch_and_wait_default_handler(RaucInstallArgs *args, gchar* bu
 			goto out;
 		}
 
-		if (dest_slot->mount_point || dest_slot->ext_mount_point) {
+		if ((dest_slot->mount_point || dest_slot->ext_mount_point) &&
+		    !(dest_slot->maybe_mounted && mfimage->hooks.install)) {
```
[the remark this answers is not in the document]
Comment (Villemoes) [reply src/install.c]: I considered simply doing the && part as an extra if statement; the only problem is that that would cost an extra level of indentation (and thus make the patch do a lot of whitespace churn). Maybe a small helper function (check_slot_updatable or something like that) would make it more readable.

[the remark this answers is not in the document]
Comment (Villemoes) [reply src/install.c]: I added an extra patch doing that refactoring; if that looks better, I can squash that with the relevant commit.

File: docs/reference.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (ohporter) [reply docs/reference.rst]: allow-mounted-updates would make it clearer.

**Comment by jluebbe** [comment]
I've rebased this to the current master. Next, I'll rename the option to config option to `allow-mounted` to make it easier to understand.

**Comment by jluebbe** [comment]
I've renamed the option to `allow-mounted` and added a check to the test that the hook was actually run.

**Comment by jluebbe** [comment]
@Villemoes what do you think of the current state of the PR?

File: src/install.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jluebbe) [reply src/install.c]: I've squashed that commit. Thanks.

File: docs/reference.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (jluebbe) [reply docs/reference.rst]: I've changed it to "allow-mounted".

**Comment by ravi-prevas** [comment]
Looks good.

**Comment by Villemoes** [comment]
Bah, I always forget to check what I'm logged in as. Just in case it's not immediately obvious, @Villemoes == @prevas-ravi .

File: test/install.c [hunk 6]
```diff
@@ -1130,6 +1151,47 @@ static void install_test_bundle_hook_post_install(InstallFixture *fixture,
...
+	args->name = g_strdup(bundlepath);
+	args->notify = install_notify;
+	args->cleanup = install_cleanup;
+	res = do_install_bundle(args, &ierror);
+	g_assert_no_error(ierror);
+	g_assert_true(res);
+
+	g_assert_true(g_file_test(hookfilepath, G_FILE_TEST_IS_REGULAR));
+
+	args->status_result = 0;
```
Comment (ejoerns) [inline test/install.c:1190]: Not really useful for anything but other tests also do this, thus maybe something to clean up separately.

File: docs/reference.rst [hunk 7]
```diff
@@ -240,6 +240,11 @@ hierarchical separator.
   active slots, so that the inactive one can be selected as the update target.
   The parent slot is referenced using the form ``<slot-class>.<idx>``.
 
+``allow-mounted=<true/false>``
+  Setting this entry ``true`` tells RAUC to that the slot may be
```
Comment (ejoerns) [inline docs/reference.rst:244]:
```suggestion
  Setting this entry ``true`` tells RAUC that the slot may be
```

File: docs/reference.rst [hunk 8]
```diff
@@ -240,6 +240,11 @@ hierarchical separator.
...
+  updated even if it is already mounted. Such a slot can only be
+  updated by a custom install hook.
```
Comment (ejoerns) [inline docs/reference.rst:246]: nit: to my understanding this should be:

```suggestion
  updated even if it is already mounted. Such a slot can be
  updated only by a custom install hook.
```

We might also do other things with the slot than updating, but this is not the point here ;)

File: test/install.c (hunk 6 above)
Comment (jluebbe) [reply test/install.c:1190]: Yes, that should be a separate PR.

**Pull request merged**